core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java

```diff
@@ -0,0 +1,109 @@
+                filterExpressionRewrite(),
+                joinExpressionRewrite(),
+                valuesExpressionRewrite(),
+                patternRecognitionExpressionRewrite());
+    }
+
+    private static Expression rewrite(Expression expression, Rule.Context context, Metadata metadata)
+    {
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Session session = context.getSession();
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(metadata, session), expression);
+    }
+
+    private static class Visitor
+            extends io.trino.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final Session session;
+
+        public Visitor(Metadata metadata, Session session)
+        {
+            this.metadata = metadata;
+            this.session = session;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            FunctionCall rewritten = treeRewriter.defaultRewrite(node, context);
+            if (metadata.decodeFunction(rewritten.getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME) &&
+                    rewritten.getArguments().size() == 1 &&
+                    rewritten.getArguments().get(0) instanceof FunctionCall &&
+                    metadata.decodeFunction(((FunctionCall) rewritten.getArguments().get(0)).getName()).getSignature().getName().equals(ArraySortFunction.NAME)) {
+                FunctionCall arraySortFunctionCall = (FunctionCall) rewritten.getArguments().get(0);
+                ResolvedFunction arraySortResolvedFunction = metadata.decodeFunction(arraySortFunctionCall.getName());
+
+                List<Expression> arraySortArguments = arraySortFunctionCall.getArguments();
+                List<Type> arraySortArgumentsTypes = arraySortResolvedFunction.getSignature().getArgumentTypes();
+
+                FunctionCall arrayDistinctCall = FunctionCallBuilder.resolve(session, metadata)
+                        .setName(QualifiedName.of(ArrayDistinctFunction.NAME))
+                        .setArguments(
+                                ImmutableList.of(arraySortArgumentsTypes.get(0)),
+                                ImmutableList.of(arraySortArguments.get(0)))
+                        .build();
+
+                FunctionCallBuilder arraySortCallBuilder = FunctionCallBuilder.resolve(session, metadata)
+                        .setName(QualifiedName.of(ArraySortFunction.NAME))
+                        .addArgument(arraySortArgumentsTypes.get(0), arrayDistinctCall);
+
+                if (arraySortArguments.size() == 2) {
+                    arraySortCallBuilder.addArgument(arraySortArgumentsTypes.get(1), arraySortArguments.get(1));
+                }
+                return treeRewriter.defaultRewrite(arraySortCallBuilder.build(), context);
```

Comment:
The rewrite is wrapped in an iterative rule, so it will be run again if there's something more to do. Let's not do the default rewrite twice in one invocation.

Reply:
Yeah, I see that this test case needs multiple passes of the rewrite to fully sort the `array_distinct` and `array_sort` calls.
What I mean by "it's wrapped in an iterative rule" is that even if a single pass doesn't do the whole job, the rule should be called again, and again... until the call chain is fully sorted.

This test runs a single invocation of the rule, so I wouldn't expect the result fully sorted.

Reply:
Thanks for the clarification, I looked at the IterativeOptimizer and indeed it will exhaustively go through a set of rules and try to apply them until. So i will adapt the test for one single invocation and remove the recursive call.